pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,628 @@
+	// Default number of workers processing a rule change.
+	defaultWorkers               = 4
+	resyncPeriod   time.Duration = 0
+)
+
+func NewEgressController(
+	kubeClient clientset.Interface,
+	ofClient openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	nodeIP net.IP,
+	egressInformer opsinformers.EgressInformer,
+	antreaClientGetter agent.AntreaClientProvider,
+	ifaceStore interfacestore.InterfaceStore,
+) *Controller {
+	c := &Controller{
+		kubeClient:           kubeClient,
+		nodeIP:               nodeIP,
+		ofClient:             ofClient,
+		ifaceStore:           ifaceStore,
+		antreaClientProvider: antreaClientGetter,
+		setByGroup:           make(map[string]v1beta1.GroupMemberSet),
+		policyMap:            make(map[string]string),
+		// snatPktMarkRange takes an 8-bit range of pkt_mark to store the ID of
+		// a SNAT IP. The bit range must match SNATIPMarkMask.
+		IPAllocator:              newIPAllocator(256),
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "egressgroup"),
+		egressPolicyInformer:     egressInformer,
+		ovsBridgeClient:          ovsBridgeClient,
+		egressPolicyLister:       egressInformer.Lister(),
+		egressPolicyListerSynced: egressInformer.Informer().HasSynced,
+		objectType:               "EgressGroup",
+		fullSynced:               false,
+	}
+
+	c.watchFunc = func() (watch.Interface, error) {
+		antreaClient, err := c.antreaClientProvider.GetAntreaClient()
+		if err != nil {
+			return nil, err
+		}
+		return antreaClient.ControlplaneV1beta1().EgressGroups().Watch(context.TODO(), metav1.ListOptions{})
+	}
+	c.AddFunc = func(obj runtime.Object) error {
+		group, ok := obj.(*v1beta1.EgressGroup)
+		if !ok {
+			return fmt.Errorf("cannot convert to *v1beta1.EgressGroup: %v", obj)
+		}
+		klog.Infof("EgressGroup %#v\n applied to Pods on this Node", group)
+		c.AddEgressGroup(group)
+		return nil
+	}
+	c.UpdateFunc = func(obj runtime.Object) error {
+		klog.Infof("egress group updating")
+		patch, ok := obj.(*v1beta1.EgressGroupPatch)
+		if !ok {
+			return fmt.Errorf("cannot convert to *v1beta1.EgressGroup: %v", obj)
+		}
+		c.PatchEgressGroup(patch)
+		return nil
+	}
+	c.DeleteFunc = func(obj runtime.Object) error {
+		klog.Infof("egress group deleting")
+		group, ok := obj.(*v1beta1.EgressGroup)
+		if !ok {
+			return fmt.Errorf("cannot convert to *v1beta1.EgressGroup: %v", obj)
+		}
+		c.DeleteEgressGroup(group)
+		return nil
+	}
+	c.ReplaceFunc = func(objs []runtime.Object) error {
+		klog.Infof("egress replacing")
+		policies := make([]*v1beta1.EgressGroup, len(objs))
+		var ok bool
+		for i := range objs {
+			policies[i], ok = objs[i].(*v1beta1.EgressGroup)
+			if !ok {
+				return fmt.Errorf("cannot convert to *v1beta1.EgressGroup: %v", objs[i])
+			}
+			klog.Infof("EgressGroup %s applied to Pods on this Node", policies[i].Name)
+		}
+		c.ReplaceEgressGroups(policies)
+		return nil
+	}
```

Comment:
The above function variables seem useless? It can be just inline code like:
```
func (c *Controller) watch() {
	klog.Info("Starting watch for EgressGroup")
	antreaClient, err := c.antreaClientProvider.GetAntreaClient()
        ...
	watcher, err := antreaClient.ControlplaneV1beta1().EgressGroups().Watch(context.TODO(), options)
        ...
        for {
		select {
		case event, ok := <-watcher.ResultChan():
			if !ok {
				return
			}
			switch event.Type {
			case watch.Added:
				klog.V(2).Infof("Added EgressGroup (%#v)", event.Object)
				c.addEgressGroup(...)
```
The abstract class in networkpolicy is to share code for 3 watchers, which doesn't make sense here, unless you extract the abstract class from networkpolicy controller to a common package, but I would defer its refactoring before other essential changes are finished.